File: app/src/main/java/dev/carloszuil/herojourney/adapter/QuestAdapter.java
package dev.carloszuil.herojourney.adapter;


import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.CheckBox;
import android.widget.ImageButton;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;
import java.util.List;

import dev.carloszuil.herojourney.R;
import dev.carloszuil.herojourney.data.local.entities.Quest;
import dev.carloszuil.herojourney.data.local.entities.QuestState;

public class QuestAdapter extends RecyclerView.Adapter<QuestAdapter.QuestViewHolder> {

    /** Callback cuando se marca/desmarca una quest */
    public interface OnQuestCheckedListener {
        void onQuestChecked(@NonNull Quest quest, boolean isChecked);
    }

    /** Callback al pulsar mover quest */
    public interface OnMoveClickedListener {
        void onMoveClicked(@NonNull Quest quest);
    }

    private final List<Quest> quests = new ArrayList<>();
    private final OnQuestCheckedListener onQuestChecked;
    private final OnMoveClickedListener onMoveClicked;

    public QuestAdapter(
            @NonNull List<Quest> initialList,
            @NonNull OnQuestCheckedListener onQuestChecked,
            @NonNull OnMoveClickedListener onMoveClicked
    ) {
        this.onQuestChecked = onQuestChecked;
        this.onMoveClicked  = onMoveClicked;
        this.quests.addAll(initialList);
    }

    public static class QuestViewHolder extends RecyclerView.ViewHolder {
        public final TextView questNombre;
        public final CheckBox questCheckBox;
        public final ImageButton btnMove;

        public QuestViewHolder(@NonNull View itemView) {
            super(itemView);
            questNombre   = itemView.findViewById(R.id.questText);
            questCheckBox = itemView.findViewById(R.id.questCheckbox);
            btnMove       = itemView.findViewById(R.id.btnMoveQuest);
        }
    }

    @NonNull
    @Override
    public QuestViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.item_quest, parent, false);
        return new QuestViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull QuestViewHolder holder, int position) {
        Quest quest = quests.get(position);
        holder.questNombre.setText(quest.getNombre());

        // Evitar disparar el listener al reciclar
        holder.questCheckBox.setOnCheckedChangeListener(null);

        // Sólo marcado si está COMPLETADA
        holder.questCheckBox.setChecked(quest.getEstado() == QuestState.QUEST_BOARD_4);

        // Reasignar listener
        holder.questCheckBox.setOnCheckedChangeListener((buttonView, isChecked) -> {
            onQuestChecked.onQuestChecked(quest, isChecked);
        });

        holder.btnMove.setOnClickListener(v -> onMoveClicked.onMoveClicked(quest));
    }

    @Override
    public int getItemCount() {
        return quests.size();
    }

    /** Actualiza la lista con DiffUtil para animaciones y cambios eficientes */
    public void updateList(@NonNull List<Quest> newList) {
        DiffUtil.DiffResult diffResult = DiffUtil.calculateDiff(
                new QuestDiffCallback(this.quests, newList)
        );
        this.quests.clear();
        this.quests.addAll(newList);
        diffResult.dispatchUpdatesTo(this);
    }
}

File: app/src/main/java/dev/carloszuil/herojourney/adapter/QuestAdapter.kt
package dev.carloszuil.herojourney.adapter

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.CheckBox
import android.widget.ImageButton
import android.widget.TextView
import androidx.cardview.widget.CardView
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import dev.carloszuil.herojourney.R
import dev.carloszuil.herojourney.model.Quest
import dev.carloszuil.herojourney.model.QuestState

class QuestAdapter(
    initialList: List<Quest>,  // Cambiar para recibir una lista inmutable
    private val onQuestChecked: (Quest, Boolean) -> Unit,
    private val onMoveClicked: (Quest) -> Unit
) : RecyclerView.Adapter<QuestAdapter.QuestViewHolder>() {

    // Crear una lista mutable interna
    private val quests = mutableListOf<Quest>().apply { addAll(initialList) }

    class QuestViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val questNombre: TextView = itemView.findViewById(R.id.questText)
        val questCheckBox: CheckBox = itemView.findViewById(R.id.questCheckbox)
        val btnMove: ImageButton = itemView.findViewById(R.id.btnMoveQuest)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): QuestViewHolder {
        val view = LayoutInflater.from(parent.context).inflate(R.layout.item_quest, parent, false)
        return QuestViewHolder(view)
    }

    override fun onBindViewHolder(holder: QuestViewHolder, position: Int) {
        val quest = quests[position]
        holder.questNombre.text = quest.nombre

        // Quitar listener antes de actualizar el checkbox
        holder.questCheckBox.setOnCheckedChangeListener(null)

        // Mostrar marcado solo si está COMPLETADA
        holder.questCheckBox.isChecked = quest.estado == QuestState.COMPLETADA

        // Reasignar listener
        holder.questCheckBox.setOnCheckedChangeListener { _, isChecked ->
            onQuestChecked(quest, isChecked)
        }

        holder.btnMove.setOnClickListener {
            onMoveClicked(quest)
        }
    }

    override fun getItemCount(): Int = quests.size

    fun updateList(newList: List<Quest>) {
        val diffResult = DiffUtil.calculateDiff(QuestDiffCallback(quests, newList))
        quests.clear()
        quests.addAll(newList)
        diffResult.dispatchUpdatesTo(this)
    }
}
